sale_loyalty_limit/tests/test_sale_loyalty_limit.py

```diff
@@ -18,241 +28,312 @@ def _create_sale(self, partner, salesman=False):
+            if "error" in status:
+                raise ValidationError(status["error"])
 
     def test_01_program_no_code_customer_limit(self):
         """A program with no code and customer application limit won't be applied
         once the limit is reached"""
         sale_1 = self._create_sale(self.partner_1)
         # In the case definition the program is no code, so there's nothing else to
         # setup.
-        sale_1.recompute_coupon_lines()
+        self._action_apply_program(sale_1, self.loyalty_promotion)
         self.assertTrue(bool(sale_1.order_line.filtered("is_reward_line")))
         # The limit is 2, so the promotion can be placed in a second order
         sale_2 = self._create_sale(self.partner_1)
-        sale_2.recompute_coupon_lines()
+        self._action_apply_program(sale_2, self.loyalty_promotion)
         self.assertTrue(bool(sale_2.order_line.filtered("is_reward_line")))
         # As we reach the limit, no discount will be applied
         sale_3 = self._create_sale(self.partner_1)
-        sale_3.recompute_coupon_lines()
+        with self.assertRaises(ValidationError):
+            self._action_apply_program(sale_3, self.loyalty_promotion)
         self.assertFalse(bool(sale_3.order_line.filtered("is_reward_line")))
         # However other partners can still enjoy the promotion
         sale_4 = self._create_sale(self.partner_2)
-        sale_4.recompute_coupon_lines()
+        self._action_apply_program(sale_4, self.loyalty_promotion)
         self.assertTrue(bool(sale_4.order_line.filtered("is_reward_line")))
 
     def test_02_program_promo_code_customer_limit(self):
         """A program with code and customer application limit will raise an error when
-        such limit is reached for a customer"""
-        self.coupon_program.promo_code_usage = "code_needed"
-        self.coupon_program.promo_code = "TEST-SALE-COUPON-LIMIT"
+        # such limit is reached for a customer"""
+        self.loyalty_promotion.program_type = "promo_code"
+        self.loyalty_promotion.rule_ids.write(
+            {"code": "TEST-SALE-COUPON-LIMIT", "rule_max_customer_application": 2}
+        )
+        self.loyalty_promotion.reward_ids.write(
+            {"discount_product_ids": self.product_a}
+        )
         # We apply it once for partner 1...
         sale_1 = self._create_sale(self.partner_1)
-        self._apply_coupon(sale_1, "TEST-SALE-COUPON-LIMIT")
+        self._apply_promo_code(sale_1, "TEST-SALE-COUPON-LIMIT")
         self.assertTrue(bool(sale_1.order_line.filtered("is_reward_line")))
+        sale_1.action_confirm()
         # We apply it twice for partner 1...
         sale_2 = self._create_sale(self.partner_1)
-        self._apply_coupon(sale_2, "TEST-SALE-COUPON-LIMIT")
+        self._apply_promo_code(sale_2, "TEST-SALE-COUPON-LIMIT")
         self.assertTrue(bool(sale_2.order_line.filtered("is_reward_line")))
+        sale_2.action_confirm()
```

Comment:
Why do we need to confirm it now?

Reply:
In this case it is not necessary, it is only necessary to generate valid coupons in promotions that generate coupons for the next sales order because if the order is not confirmed it generates the coupon but it will not be valid until it is confirmed. I have made the change in the tests so that they are only confirmed in these cases.